File: packages/tonic-core/src/hydration/runtime.ts
import { ChromaVectorIndex } from "./chromaVectorIndex";
import { probeHydrationChromaHeartbeat, type HydrationChromaHeartbeat } from "./chromaClient";
import { HydrationRuntimeError } from "./errors";
import { MemoryVectorIndex } from "./memoryVectorIndex";
import { resolveHydrationRuntimeConfig, runtimeModeToBackend, type HydrationRuntimeConfig } from "./runtimeConfig";
import type { VectorIndex } from "./vectorIndex";

export type HydrationRuntime = {
  backend: ReturnType<typeof runtimeModeToBackend>;
  index: VectorIndex;
  config: HydrationRuntimeConfig;
};

export async function probeHydrationRuntimeReadiness(
  repoRoot: string,
  env: NodeJS.ProcessEnv = process.env,
  overrides: Partial<HydrationRuntimeConfig> = {},
): Promise<HydrationChromaHeartbeat | null> {
  const config = resolveHydrationRuntimeConfig(repoRoot, env, overrides);
  if (config.mode !== "http") {
    return null;
  }
  return probeHydrationChromaHeartbeat(config);
}

export async function createHydrationRuntime(
  repoRoot: string,
  env: NodeJS.ProcessEnv = process.env,
  overrides: Partial<HydrationRuntimeConfig> = {},
): Promise<HydrationRuntime> {
  const config = resolveHydrationRuntimeConfig(repoRoot, env, overrides);
  const backend = runtimeModeToBackend(config.mode);
  if (config.mode === "memory") {
    return {
      backend,
      config,
      index: new MemoryVectorIndex(config.collectionName),
    };
  }
  if (config.mode !== "http") {
    throw new HydrationRuntimeError(
      `TypeScript hydration currently supports Chroma only through HTTP mode. Received TONIC_CHROMA_MODE=${config.mode}.`,
    );
  }
  return {
    backend,
    config,
    index: new ChromaVectorIndex(config),
  };
}

File: merge-tonic-lib/tonic/hydration/runtime.py
"""Hydration runtime factory for vector backends."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

from .chroma_client import HydrationChromaHeartbeat, probe_hydration_chroma_heartbeat
from .chroma_vector_index import ChromaVectorIndex
from .memory_vector_index import MemoryVectorIndex
from .runtime_config import (
    HydrationRuntimeConfig,
    resolve_hydration_runtime_config,
    runtime_mode_to_backend,
)


class SupportsVectorIndex(Protocol):
    @property
    def backend(self) -> str: ...

    @property
    def collection_name(self) -> str: ...


@dataclass(frozen=True)
class HydrationRuntime:
    backend: str
    config: HydrationRuntimeConfig
    index: SupportsVectorIndex


def create_hydration_runtime(
    repo_root: str,
    env: dict[str, str] | None = None,
    overrides: dict[str, str] | None = None,
) -> HydrationRuntime:
    config = resolve_hydration_runtime_config(repo_root, env, overrides)
    backend = runtime_mode_to_backend(config.mode)
    if config.mode == "memory":
        return HydrationRuntime(
            backend=backend,
            config=config,
            index=MemoryVectorIndex(config.collection_name),
        )
    return HydrationRuntime(
        backend=backend,
        config=config,
        index=ChromaVectorIndex(config),
    )


def probe_hydration_runtime_readiness(
    repo_root: str,
    env: dict[str, str] | None = None,
    overrides: dict[str, str] | None = None,
) -> HydrationChromaHeartbeat | None:
    config = resolve_hydration_runtime_config(repo_root, env, overrides)
    if config.mode != "http":
        return None
    return probe_hydration_chroma_heartbeat(config)
